fix: find correct msbuild executable

aniruddhadave opened the pull request:
## Description
* Use VS installation paths to find the correct msbuild path that have the required targets.

---
*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,40 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
+                {
+                    StartInfo = new ProcessStartInfo
+                    {
+                        FileName = Path.Combine(projectDirectory, "vswhere.exe"),
+                        Arguments = "-latest -requires Microsoft.Component.MSBuild -find MSBuild\\**\\Bin\\MSBuild.exe",
+                        UseShellExecute = false,
+                        RedirectStandardOutput = true,
+                        CreateNoWindow = true,
+                        WorkingDirectory = ""
```
Where does vswhere.exe come from? Is this going to be packaged with codelyzer?

aniruddhadave replied:
Yes. It is in the src/ directory. We can change the location to some other directory if required.

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,40 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
+                var msbuildExe = "";
+                while (!proc.StandardOutput.EndOfStream)
+                {
+                    msbuildExe = proc.StandardOutput.ReadLine();
```
Are we always sure it's a one line output?

aniruddhadave replied:
Yes, it should always give the lastest vs installation in one line or nothing at all which will be covered in the next catch block.

marknfawaz commented:
One additional issue here is that we're not only looking to find _any_ msbuild. Sometimes we might need to skip one we found because it's missing some targets

aniruddhadave commented:
> One additional issue here is that we're not only looking to find _any_ msbuild. Sometimes we might need to skip one we found because it's missing some targets

Is there some other kind of logic to it? If the logic is to only find the latest sdk/vs installation then it should be covered by these changes. However, if there is some other logic to find the correct msbuild we would need to incorporate it.

saikiranakula-amzn commented:
Tests are failing

saikiranakula-amzn commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -528,5 +549,71 @@ public void Dispose()
+
+                var msBuild2022 = Path.Combine(vsPath2022, "MSBuild.exe");
+                var msBuild2019 = Path.Combine(vsPath2019, "MSBuild.exe");
+                var msBuild2017 = Path.Combine(vsPath2017, "MSBuild.exe");
```
Not major one but some people might still use vs2015, good to add the path for vs2015 as well

aniruddhadave replied:
msbuild 14 is the path for VS 2015.

aniruddhadave commented:
> 
> 
> Tests are failing

The test for verifying the number of dlls in SampleWebApi was failing. The number of dlls on my local machine are 15 due to which I updated the test case, however, on codebuild the number of dlls is 16. We need to verify this behavior for the test case. For now I have reverted back to the original.

aniruddhadave commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,25 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
+
             options.EnvironmentVariables.Add(Constants.EnableNuGetPackageRestore, Boolean.TrueString.ToLower());
 
             options.Arguments.Add(Constants.RestorePackagesConfigArgument);
```
Buildalyzer uses the Msbuild path from the environment variable `MSBUILD_EXE_PATH`
https://github.com/daveaglick/Buildalyzer/blob/e30d44755ad17e8abf3efb53acc4a5d4f5a77ea7/src/Buildalyzer/Environment/BuildEnvironment.cs#L67

aniruddhadave commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,25 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
+            {
+                var msbuildExe = "";
+                var editions = new string[] { "Enterprise", "Professional", "Community" };
+                var targets = new string[] { "Microsoft.CSharp.targets", "Microsoft.CSharp.CurrentVersion.targets", "Microsoft.CSharp.CrossTargeting.targets" };
```
I wasn't able to find a solid source online to verify this. I did do a cross-check with Mark about it. It would be helpful if I can get more suggestions on verifying this.

saikiranakula-amzn replied:
I will do a quick check as well :)

aniruddhadave commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -528,5 +549,71 @@ public void Dispose()
+            var vspath14 = @"msbuild\14.0\bin\";
+            var vspath12 = @"msbuild\12.0\bin\msbuild.exe";
+
+            var msbuild14 = Path.Combine(vspath14, "msbuild.exe");
```
Msbuild 16.0 is the same as Visual Studio 2019 and Msbuild 17.0 is the same as VS 2022
https://docs.microsoft.com/en-us/visualstudio/msbuild/whats-new-msbuild-16-0?view=vs-2022
https://docs.microsoft.com/en-us/visualstudio/msbuild/whats-new-msbuild-17-0?view=vs-2022

saikiranakula-amzn commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,25 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
+            }
+            catch
+            {
+                var error = "No instances of MSBuild could be detected.";
```
Log this message

birojnayak commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,25 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
+                var targets = new string[] { "Microsoft.CSharp.targets", "Microsoft.CSharp.CurrentVersion.targets", "Microsoft.CSharp.CrossTargeting.targets" };
+
+                msbuildExe = GetLatestVSMSBuildPath();
+                Environment.SetEnvironmentVariable("MSBUILD_EXE_PATH", msbuildExe);
```
this should be EnvironmentVariable just as used before

birojnayak commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,25 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
 
+            try
+            {
+                MSBuildLocator.RegisterDefaults();
```
this is Windows only API.... had sync with Mark, we are getting rid of buildlocator... so I am assuming we are not using this API anymore...

aniruddhadave replied:
Yes we are not going to use this anymore.

aniruddhadave commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -528,5 +549,71 @@ public void Dispose()
+            var editions = new string[] { "Enterprise", "Professional", "Community" };
+            var targets = new string[] { "Microsoft.CSharp.targets", "Microsoft.CSharp.CurrentVersion.targets", "Microsoft.CSharp.CrossTargeting.targets" };
+
+            foreach (string editionType in editions)
```
This logic orders the files by the last time written. So if an older version is installed after a new one it will pick up the older one.
https://github.com/daveaglick/Buildalyzer/blob/da88770a2952f7894c05cd0866ef9d766f78aaea/src/Buildalyzer/Environment/EnvironmentFactory.cs#L174

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -528,5 +541,116 @@ public void Dispose()
+
+        public static string GetVersionFolder(string vsPath)
+        {
+            List<string> elements = vsPath.Split(@"\").ToList();
```
You can use Path.DirectorySeparatorChar instead to make sure it's cross-platform

marknfawaz replied:
Same as above. You can use DirectorySeparatorChar

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -528,5 +541,116 @@ public void Dispose()
+            string[] elements = vsPath.ToLower().Split(@"\");
+            foreach (var edition in editions)
+            {
+                if (elements.Contains(edition.ToLower()))
```
nit: Instead of ToLower() on the edition, why not save it that way?

marknfawaz commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -528,5 +541,116 @@ public void Dispose()
+                    })
+                    .Where(msbuild =>
+                    {
+                        var targetsWithPath = GetTargetsWithPath(msbuild.DirectoryName, targets);
+                        if (targetsWithPath.TrueForAll(File.Exists)) return true;
```
Did you check if the same targets exist for lower versions?

aniruddhadave replied:
Yes they exist for MSBuild versions >= 14.0

marknfawaz reviewed (changes requested):
Looks good. I added some minor comments.. Let me know your thoughts

saikiranakula-amzn commented on src/Analysis/Codelyzer.Analysis.Build/WorkspaceBuilderHelper.cs:
```diff
@@ -459,6 +461,17 @@ private EnvironmentOptions GetEnvironmentOptions(IProjectFile projectFile)
+            {
+                var msbuildExe = "";
+                msbuildExe = GetFrameworkMsBuildExePath();
+                options.EnvironmentVariables.Add(EnvironmentVariables.MSBUILD_EXE_PATH, msbuildExe);
```
minor check here. If msbuildExe is not found. Dont set the environment variable
